File: javascript/arrays/Merge.js
/**
 * You are given two integer arrays nums1 and nums2, sorted in non-decreasing order,
 * and two integers m and n, representing the number of elements in nums1 and nums2
 * respectively.Merge nums1 and nums2 into a single array sorted in non-decreasing
 * order. The final sorted array should not be returned by the function, but instead
 * be stored inside the array nums1. To accommodate this, nums1 has a length of m + n,
 * where the first m elements denote the elements that should be merged, and the
 * last n elements are set to 0 and should be ignored. nums2 has a length of n.
 *
 * Example 1:
 * Input: nums1 = [1,2,3,0,0,0], m = 3, nums2 = [2,5,6], n = 3
 * Output: [1,2,2,3,5,6]
 * Explanation: The arrays we are merging are [1,2,3] and [2,5,6]. The result of
 * the merge is [1,2,2,3,5,6] with the underlined elements coming from nums1.
 *
 * Constraints:
 * nums1.length == m + n
 * nums2.length == n
 * 0 <= m, n <= 200
 * 1 <= m + n <= 200
 * -10^9 <= nums1[i], nums2[j] <= 10^9
 */

/**
 * 同时对两数组从右往左扫描，取扫描到的元素中较大的那个依次放到目标数组中，当nums1
 * 排好序以后，只需要将nums2剩下的放到nums1中即可
 */
var merge = function(nums1, m, nums2, n) {
    let i = m - 1;
    let j = n - 1;
    let k = m + n - 1;

    while (i >= 0 && j >= 0) {
        if (nums1[i] > nums2[j]) {
            nums1[k] = nums1[i];
            i--;
        }
        else {
            nums1[k] = nums2[j];
            j--;
        }
        k--;
    }
    while (j >= 0) {
        nums1[k] = nums2[j];
        k--;
        j--;
    }
};

let nums1 = [1,2,3,0,0,0];
let nums2 = [2,5,6];
let m = 3;
let n = 3;
console.log('input array1: ' + nums1);
console.log('input array2: ' + nums2);
console.log('m: ' + m);
console.log('n: ' + n);
merge(nums1, m, nums2, n);
console.log('output array1: ' + nums1);

File: java/sort/Merge.java
/**
 * Given an array of intervals where intervals[i] = [start_i, end_i], merge all overlapping
 * intervals, and return an array of the non-overlapping intervals that cover all
 * the intervals in the input.
 *
 * Example 1:
 * Input: intervals = [[1,3],[2,6],[8,10],[15,18]]
 * Output: [[1,6],[8,10],[15,18]]
 * Explanation: Since intervals [1,3] and [2,6] overlap, merge them into [1,6].
 * 
 * Constraints:
 * 1 <= intervals.length <= 10^4
 * intervals[i].length == 2
 * 0 <= start_i <= end_i <= 10^4
 */

package sort;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;
import arrays.ArrayUtils;

public class Merge {

    public static void main(String[] args) {
        int[][] intervals = {{1,3},{2,6},{8,10},{15,18}};
        ArrayUtils a = new ArrayUtils();
        a.print2dArray(intervals);
        Merge m = new Merge();
        a.print2dArray(m.merge(intervals));
    }

    /**
     * 在原数组上进行修改，首先进行排序，然后从左往右遍历数组，每次遇到一个间隔，就
     * 比较当前间隔的起点和上一个间隔的终点，如果前者小于等于后者，则选择两个间隔终
     * 点较大的作为新的终点，否则添加当前间隔到结果中
     */
    public int[][] merge(int[][] intervals) {
        if (intervals.length <= 1) {
            return intervals;
        }
        Arrays.sort(intervals, (a, b) -> (a[0] - b[0]));
        int[] pre = null;
        int i = 0;
        for (int[] cur : intervals) {
            if (pre != null && pre[1] >= cur[0]) {
                pre[1] = (pre[1] >= cur[1] ? pre[1] : cur[1]);
            } else {
                intervals[i] = cur;
                i++;
                pre = cur;
            }
        }
        return i == intervals.length ? intervals : Arrays.copyOfRange(intervals, 0, i);
    }

}
